backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt

```diff
@@ -432,10 +429,15 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
                                     } else {
                                         null
                                     }
-                                    if (irField.type.binaryTypeIsReference()) {
-                                        call(context.llvm.initGlobalFunction, listOf(address, initialValue ?: kNullObjHeaderPtr))
-                                    } else if(initialValue != null) {
-                                        store(initialValue, address)
+                                    val needRegistration =
+                                            context.memoryModel == MemoryModel.EXPERIMENTAL && // only for the new MM
+                                                    irField.type.binaryTypeIsReference() && // only for references
+                                                    (initialValue != null || // which are initialized from heap object
+                                                            !irField.isFinal) // or are not final
+                                    if (needRegistration) {
+                                        call(context.llvm.initAndRegisterGlobalFunction, listOf(address))
```

Comment:
Shouldn't the argument be `listOf(address, initialValue)`?

Reply:
Right, thanks!